source/client_channel_handler.c

```diff
@@ -94,11 +89,46 @@ static void s_on_time_to_ping(struct aws_channel_task *channel_task, void *arg,
         s_schedule_ping(connection);
     }
 }
-static int s_packet_handler_connack(
+
+static int s_validate_received_packet_type(
     struct aws_mqtt_client_connection_311_impl *connection,
-    struct aws_byte_cursor message_cursor) {
+    enum aws_mqtt_packet_type packet_type) {
+    { /* BEGIN CRITICAL SECTION */
+        mqtt_connection_lock_synced_data(connection);
+        /* [MQTT-3.2.0-1] The first packet sent from the Server to the Client MUST be a CONNACK Packet */
+        if (connection->synced_data.state == AWS_MQTT_CLIENT_STATE_CONNECTING &&
+            packet_type != AWS_MQTT_PACKET_CONNACK) {
+            mqtt_connection_unlock_synced_data(connection);
+            AWS_LOGF_ERROR(
+                AWS_LS_MQTT_CLIENT,
+                "id=%p: First message received from the server was not a CONNACK. Terminating connection.",
+                (void *)connection);
+            aws_channel_shutdown(connection->slot->channel, AWS_ERROR_MQTT_PROTOCOL_ERROR);
+            return aws_raise_error(AWS_ERROR_MQTT_PROTOCOL_ERROR);
+        }
+        mqtt_connection_unlock_synced_data(connection);
+    } /* END CRITICAL SECTION */
 
-    AWS_LOGF_TRACE(AWS_LS_MQTT_CLIENT, "id=%p: CONNACK received", (void *)connection);
+    if (AWS_UNLIKELY(packet_type > AWS_MQTT_PACKET_DISCONNECT || packet_type < AWS_MQTT_PACKET_CONNECT)) {
```

Comment:
shouldn't this check be in the decoder, since the callback we invoke is based on packet_type? so an unexpected packet_type would be an invalid array access?

Reply:
The safe packet handler wrapper checks for null.